# 0. przypomnienie z poprzednich warsztatów - obsługa jupyter notebook   
**edycja - podświetlenie zielone**   
`shift`+`enter` - wywołanie komórki i przejście do kolejnej (stworzenie kolejnej)    
`ctrl`+`enter` - wywołanie komórki   

**wybór - podświetlenie niebieskie**    
`A` - stworzenie komórki nad obecną komórką (Above)       
`B` - stworzenie komórki pod obecną komórką (Below)    
`D`+`D` - usunięcie podświetlonej komórki

**gdy nie wiemy jak działa dana funkcja w bibliotece**    
`shift`+`tab` - wyświetlenie dokumentacji (DocString)

## Forma Warsztatów
* Blok numpy 50 minut
* Przerwa 15 minut 
* Blok pandas 55 minut

Każdy blok ma w środku 5 minutowe zadanko i jest zakończony 2 zadaniami których zrobienie powinno zająć max 15 minut. 

In [1]:
import numpy as np 

# 1. Wprowadzenie do Numpy 
**Numpy** - **Num**erical **py**thon jest to jedna z najważniejszych bibliotek do obliczeń numerycznych w pythonie. Używamy jej bo :    
1. jest szybka
2. jest lekka
3. jest wygodna

## 1.1 Pokażemy teraz szybkość Numpy

In [2]:
import time
import sys

In [3]:
python_array = [i for i in range(1000000)] # milion elementów od 0 do 999999
numpy_array  = np.arange(1000000)

In [4]:
# obliczmy średnią z list
# python
sum_of_elements = 0 
start1 = time.time()

for element in python_array :
    sum_of_elements += element 
    
mean_python = sum_of_elements/len(python_array)

end1 = time.time()

In [5]:
# numpy 
start2 = time.time()

mean_numpy =  numpy_array.mean()

end2 = time.time()

In [6]:
# porównanie
time_python = end1 - start1
time_numpy  = end2 - start2
print(f'Średnie wynoszą odpowiednio {mean_python} i {mean_numpy}, obliczenia zajęły odpowiednio {round(time_python,4)} i {round(time_numpy,4)} sekund. \n')
print(f'W tych obliczeniach Numpy był {round(time_python/time_numpy)} razy szybszy. Przyjmuje się, że w zależności od obliczeń numpy jest od 10 do 100 razy szybszy')

Średnie wynoszą odpowiednio 499999.5 i 499999.5, obliczenia zajęły odpowiednio 0.0656 i 0.001 sekund. 

W tych obliczeniach Numpy był 65 razy szybszy. Przyjmuje się, że w zależności od obliczeń numpy jest od 10 do 100 razy szybszy


## 1.2 Pokażemy, że obiekty w Numpy zajmują mniej miejsca

Numpy korzysta z dobrze zaimplementowanego (w C) zarządzania pamięcią. Dodatkowo każda lista ma jeden typ danych, w przeciwieństwie do pythona, w którym każdy element może być innego typu. 

In [7]:
numpy_array = numpy_array.astype('int32') # domyślnie może się różnić 

In [8]:
print(f'python "waży" {sys.getsizeof(python_array)/(1048576)} MB')
print(f'numpy  "waży" {sys.getsizeof(numpy_array)/(1048576)} MB')

python "waży" 8.294540405273438 MB
numpy  "waży" 3.814788818359375 MB


## 1.3 Wygodę postaramy się udowodnić w ciągu tego warsztatu! 

# 2. Podstawy 

In [9]:
import numpy as np

## 2.1 Typy danych w listach numpy

In [10]:
# listę w pythonie tworzymy np tak: 
[1,2,3,4,5]

[1, 2, 3, 4, 5]

In [11]:
# stwórzmy array listę z listy pythonowej
np.array([1,2,3,4,5])

array([1, 2, 3, 4, 5])

In [12]:
a = np.array([1,2,3,4,5])
a

array([1, 2, 3, 4, 5])

In [13]:
type(a)

numpy.ndarray

In [14]:
# typ listy zależny od tego co wrzucimy do listy pythonowej
a.dtype

dtype('int64')

`dtype` - typ skalarów w liście. Jedna lista może mieć jeden skalar

In [15]:
# możemy inicjować array listy z wyznaczonym przez nas typem
np.array([1,2,3,4,5], dtype = "float32")

array([1., 2., 3., 4., 5.], dtype=float32)

In [16]:
np.array([1,2,3,4,5], dtype = "str")  # zwraca unicode

array(['1', '2', '3', '4', '5'], dtype='<U1')

### Hierarchia dtype

In [17]:
# co się stanie jak floaty znajdą się w liście intów?
np.array([1,2,3.14,2.91313213213], dtype = "int64")

array([1, 2, 3, 2])

In [18]:
# a co jeśli w liście znajdzie się string (napis)? 
np.array([1,2,3,"jeleń"])

array(['1', '2', '3', 'jeleń'], dtype='<U21')

In [19]:
# a co jeśli zmienimy dtype listy na int?
np.array([1,2,3,"jeleń"], dtype = "int64")

ValueError: invalid literal for int() with base 10: 'jeleń'

## 2.2 tworzenie list w numpy

Możemy też korzystać z wbudowanych "generatorów" list, czyli metod obiektu np. 
Tworzenie takich list jest podobne co wyrażeń listotwórczych za pomocą `[for i in range(...)]` w pythonie. 
Bardzo bliska w działaniu jest metoda `arange`

In [22]:
[i for i in range(5)]

[0, 1, 2, 3, 4]

In [23]:
np.arange(5) # shift + tab daje nam docstringa

array([0, 1, 2, 3, 4])

In [24]:
# co zwróci? 
[i for i in range(3,10,2)]

[3, 5, 7, 9]

In [25]:
np.arange(3,10,2)

array([3, 5, 7, 9])

In [26]:
# możemy też tworzyć specyficzne, bardzo często przydatne listy 
np.zeros(10)   # tworzy 10 zer

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [27]:
np.ones(10)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [28]:
# ważne! repeat - często przydatne a nierozumiane
np.repeat(3,5)

array([3, 3, 3, 3, 3])

In [29]:
np.repeat([3,4], 4)  # jak to rozumieć? 

array([3, 3, 3, 3, 4, 4, 4, 4])

In [30]:
np.array([3,4]).repeat(2).repeat(2)

array([3, 3, 3, 3, 4, 4, 4, 4])

In [31]:
np.array([3,4]).repeat([1,3])

array([3, 4, 4, 4])

In [32]:
# bardziej intuicyjne
np.tile([3,4],3)

array([3, 4, 3, 4, 3, 4])

## 2.3 ndarray - wielowymiarowość

W Numpy jednym z najważniejszych obiektów jest `ndarray`. Jest to wielowymiarowa lista o jednym typie danych (skalarów). 

In [33]:
type(np.ones(10))

numpy.ndarray

Wszystkie nasze poprzednie przykłady były dla list jednowymiarowych, teraz zobaczymy jak się operuje na listach w wielu wymiarach

In [34]:
a = [[1,2,3],
     [1,2,3]]

b = np.array(a)
b

array([[1, 2, 3],
       [1, 2, 3]])

In [35]:
# wymiar
b.ndim

2

In [36]:
# kształt - ile wierszy, kolumn
b.shape

(2, 3)

In [37]:
# możemy zmieniać kształt
c = np.ones(24, dtype = "int32")
c

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1], dtype=int32)

In [38]:
c.shape

(24,)

In [39]:
# możemy zmienić kształt naszej listy
c.reshape(24,1)

array([[1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1]], dtype=int32)

In [40]:
c.reshape(4,6)

array([[1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1]], dtype=int32)

**Uwaga!** możemy wykonywać reshape, dopóki nasz shape pomieści dokładnie tyle danych ile ma elementów pierwotna lista.

In [41]:
# W tym przypadku shape to dzielniki 24, które pomnożone przez siebie dają 24, czyli ilość naszych elementów
c = c.reshape(8,3)
c

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]], dtype=int32)

In [42]:
# gdy nie spełnimy tego warunku : 
c.reshape(4,5)

ValueError: cannot reshape array of size 24 into shape (4,5)

In [43]:
# wszystko do 1 wymiaru
d = c.flatten()
d   

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1], dtype=int32)

In [44]:
# na 3 wymiarach się nie kończy, możemy iść coraz wyżej pamiętając o powyższej zasadzie
d.reshape(4,3,2) # lista mająca 4 listy, które mają 3 listy po 2 elementy

array([[[1, 1],
        [1, 1],
        [1, 1]],

       [[1, 1],
        [1, 1],
        [1, 1]],

       [[1, 1],
        [1, 1],
        [1, 1]],

       [[1, 1],
        [1, 1],
        [1, 1]]], dtype=int32)

In [45]:
d.reshape(2,3,2,2) # lista mająca 2 listy, która ma jako elementy 2 listy, które mają 3 listy, które mają 2 elementy (skalary)

array([[[[1, 1],
         [1, 1]],

        [[1, 1],
         [1, 1]],

        [[1, 1],
         [1, 1]]],


       [[[1, 1],
         [1, 1]],

        [[1, 1],
         [1, 1]],

        [[1, 1],
         [1, 1]]]], dtype=int32)

### Kształt a oś

![Shapes](img/numpy_array_t.png)

Bardzo ważny obrazek, należy z niego zapamiętać nie tylko jak działają kształty, ale również osie (**axis**). Będziemy z tego korzystać!

## 2.4 Indeksowanie

Przypuśćmy, że mamy listę n wymiarową a. Aby dostać się do naszych danych będziemy indeksować w ten sposób:    
`x[exp1, exp2]` - gdzie exp1,exp2 to slice typu `start:stop:step`, które odpowiadają odpowiednim wymiarom.

In [46]:
# weźmy array lista 
a = np.arange(20).reshape(4,5)
a

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

In [47]:
# co tu dostaniemy? 
a[2]

array([10, 11, 12, 13, 14])

In [48]:
a[1:3]

array([[ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [49]:
a[1:3,2:4]

array([[ 7,  8],
       [12, 13]])

In [50]:
a[:,1:5]

array([[ 1,  2,  3,  4],
       [ 6,  7,  8,  9],
       [11, 12, 13, 14],
       [16, 17, 18, 19]])

In [51]:
# możemy też indeksować logicznie
a > 10 

array([[False, False, False, False, False],
       [False, False, False, False, False],
       [False,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True]])

In [52]:
a[a>10]

array([11, 12, 13, 14, 15, 16, 17, 18, 19])

In [53]:
# co teraz dostaniemy? 
(a>3) and (a<10)         

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [54]:
a>3 & a<10 

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [55]:
# co musimy zrobić 
(a>3) & (a<10)      # nie zadziała pythonowy and oraz or, zamienimy je na & oraz |
                    # nawiasy są kluczowe
                    # 

array([[False, False, False, False,  True],
       [ True,  True,  True,  True,  True],
       [False, False, False, False, False],
       [False, False, False, False, False]])

In [56]:
# A co teraz?
a[(a>3)&(a<10)]

array([4, 5, 6, 7, 8, 9])

In [57]:
a[(a>18)|(a<3)]

array([ 0,  1,  2, 19])

In [58]:
# jeszcze raz dla przypomnienia
a

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

In [59]:
# znajdowanie indeksów
index_array = np.where((a > 3) & (a < 10))
index_array 

(array([0, 1, 1, 1, 1, 1]), array([4, 0, 1, 2, 3, 4]))

In [60]:
a[index_array]

array([4, 5, 6, 7, 8, 9])

## 2.5 Zadanie (*)   (5 min)

1) Stwórz macierz A taką, że:   

A = $\begin{bmatrix}
1 & 2 & 3 & 4 & 5\\
6 & 7 & 8 & 9 & 10\\
11 & 12 & 13 & 14 & 15
\end{bmatrix}$

2) spróbuj indeksami dostać się do elementów:    
  ![Pic1](img/zad1_2.png)

3)     
  ![Pic1](img/zad1_3.png)

4)     
  ![Pic1](img/zad1_1.png)
 
5) znajdź indeksy wartości mniejszych niż 10, ale większych od 4 

# 3. Operacje na listach
## 3.1 Podstawowe działania

In [62]:
# weźmy listy a,b 
a = np.array([1,1,2,1])
b = np.array([1,2,3,4])

In [63]:
a + 3

array([4, 4, 5, 4])

In [64]:
a * 3

array([3, 3, 6, 3])

In [65]:
a + b

array([2, 3, 5, 5])

In [66]:
a * b 

array([1, 2, 6, 4])

In [67]:
a - b

array([ 0, -1, -1, -3])

In [68]:
a/b

array([1.        , 0.5       , 0.66666667, 0.25      ])

In [69]:
a**b

array([1, 1, 8, 1])

In [70]:
# co jesli listy nie są tych samych kształtów 
a = np.arange(5)
b = np.repeat(2,4)

print(f"a.shape = {a.shape} \nb.shape = {b.shape}")

a.shape = (5,) 
b.shape = (4,)


### Uwaga! Musimy pilnować kształtu przy zwektoryzowanych operacjach

In [71]:
a + b 

ValueError: operands could not be broadcast together with shapes (5,) (4,) 

## 3.2 Metody obiektów ndarray + czyli kilka przydatnych rzeczy :) 

Obiekty klasy ndarray ma zdefiniowane różne metody, takie jak średnia, mediana, suma itp... Wykonujemy te operacje tak jak metody w Python. Na przykład `a.sum()`, `a.mean()`, zwrócą nam jedną liczbę. Gdy chcemy na przykład zsumować kolumy lub wiersze, będziemy się posługiwali się `axis`. 

In [72]:
# jeden wymiar 
a = np.arange(10,21)
a

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20])

In [73]:
# średnia
a.mean()

15.0

In [74]:
# suma 
a.sum()

165

In [75]:
# max, minimum analogicznie
a.max()

20

In [76]:
# indeks maksymalnej wartości
a.argmax()

10

In [77]:
# czy ten sam? 
a[a.argmax()]

20

In [78]:
a.cumsum()

array([ 10,  21,  33,  46,  60,  75,  91, 108, 126, 145, 165])

![Shapes2](img/numpy_array_t.png)

In [79]:
# mając taką listę
a

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20])

In [80]:
# domyślnie dla 1 wymiaru axis ustawiony jest na 0
a.sum(axis = 0 )

165

In [81]:
# operacje na macierzach 

A = np.arange(5*5).reshape(5,5) # 5x5
A

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [82]:
# suma wszystkich wartości
A.sum()

300

In [83]:
A.sum(axis = 0)

array([50, 55, 60, 65, 70])

In [84]:
A.sum(axis = 1)

array([ 10,  35,  60,  85, 110])

In [85]:
# co z tego wyjdzie?
A.sum(axis = 2)

AxisError: axis 2 is out of bounds for array of dimension 2

In [86]:
# dla przypomnienia
A

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [87]:
A.max(axis = 0)

array([20, 21, 22, 23, 24])

In [88]:
A.max(axis = 1)

array([ 4,  9, 14, 19, 24])

In [89]:
# czyli na przykład możemy wykonywać takie obliczenia: 
b = np.array([1,-1,1,-1,1])
b

array([ 1, -1,  1, -1,  1])

In [90]:
# czy możemy pomnożyć? 
A * b 

array([[  0,  -1,   2,  -3,   4],
       [  5,  -6,   7,  -8,   9],
       [ 10, -11,  12, -13,  14],
       [ 15, -16,  17, -18,  19],
       [ 20, -21,  22, -23,  24]])

In [91]:
print(f'A.shape = {A.shape}, b.shape = {b.shape}.') 

A.shape = (5, 5), b.shape = (5,).


In [92]:
b.reshape(5,1)

array([[ 1],
       [-1],
       [ 1],
       [-1],
       [ 1]])

In [93]:
# jak pomnożyć wiersze razy b ? 
A * b.reshape(5,1)

array([[  0,   1,   2,   3,   4],
       [ -5,  -6,  -7,  -8,  -9],
       [ 10,  11,  12,  13,  14],
       [-15, -16, -17, -18, -19],
       [ 20,  21,  22,  23,  24]])

In [94]:
# jak pomnożyć sumy kolumn przez b? 
A.sum(axis = 0) * b

array([ 50, -55,  60, -65,  70])

In [95]:
# tak samo możemy zrobić z rzędami, bo A degraduje się o jeden wymiar
A.sum(axis = 1).shape

(5,)

In [96]:
B = np.arange(50).reshape(5,10)
B

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49]])

In [97]:
A

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [98]:
# algebraiczne mnożenie macierzy
A @ B

array([[ 300,  310,  320,  330,  340,  350,  360,  370,  380,  390],
       [ 800,  835,  870,  905,  940,  975, 1010, 1045, 1080, 1115],
       [1300, 1360, 1420, 1480, 1540, 1600, 1660, 1720, 1780, 1840],
       [1800, 1885, 1970, 2055, 2140, 2225, 2310, 2395, 2480, 2565],
       [2300, 2410, 2520, 2630, 2740, 2850, 2960, 3070, 3180, 3290]])

In [99]:
# lub 
np.dot(A,B)

array([[ 300,  310,  320,  330,  340,  350,  360,  370,  380,  390],
       [ 800,  835,  870,  905,  940,  975, 1010, 1045, 1080, 1115],
       [1300, 1360, 1420, 1480, 1540, 1600, 1660, 1720, 1780, 1840],
       [1800, 1885, 1970, 2055, 2140, 2225, 2310, 2395, 2480, 2565],
       [2300, 2410, 2520, 2630, 2740, 2850, 2960, 3070, 3180, 3290]])

## 3.3 zadania  (15 min)

### Zadanie 1
Mając funcję daną wzorem   

$f(a,b) = \frac{(a^2+b^3)^b - cos( \pi * a)}{10}$       

uzupełnij funckję `dzialanie`, która zwraca wartość funkcji dla a i b. Pamiętaj o sprawdzeniu poprawności wprowadzanych danych. Jeźeli dane nie są poprawne zwroc komunikat "Nieprawidlowy ksztalt"

In [100]:
def dzialanie(a,b): 
    """
    a,b - ndarray
    out - zwracana lista, lub komunikat o błędzie 
    """
    # podpowiedź: jako funkcji cos() użyj np.cos(ndarray), a jako pi np.pi
    # do badania kształtu użyj konstrukcji warunkowych if, else
    
    #=======================================Twoj kod=======================================



    #======================================================================================
    
    return out

In [101]:
# nie zmieniać kodu!
a = np.array([3,2,1])
b = np.array([1,1,2])

dzialanie(a,b)

array([1.1, 0.4, 8.2])

Odpowiedz:  [1.1, 0.4, 8.2]

### Zadanie 2 
Majac funkcje `dzialanie2` uzupelnij kod tak, aby przy danych listach A i A (o wymiarach (20,)) konwertowal je do macierzy która ma 4 rzędy i 5 kolumn, a nastepnie stworzyc macierz ktora powstala w nastepujacy sposob: 
1. Kolumna `a` o indeksie 2 jest mnożna przez sumę wiersza `b` o indeksie 3. 
2. Kolumna `a` o indeksie 3 zamieniona jest przez kolumne `b` o indeksie 4 pomnozona przez -1.
3. Macierz powstałą z kroków 1 oraz 2 nazywamy C.
4. Co drugi wiersz macierzy (poczynając od pierwszego) jest mnożony przez 2 

In [102]:
def dzialanie2(A,B): 
    """
    A, B - 1d arrays
    """
    #===============================================Twój kod======================================================

    
    #================================================koniec========================================================
    
    return C

In [103]:
# nie zmieniać kodu!
A = np.arange(1,21)
B = np.arange(20,40)

dzialanie2(A,B)

array([[   2,    4, 1110,  -48,   10],
       [   6,    7, 1480,  -29,   10],
       [  22,   24, 4810,  -68,   30],
       [  16,   17, 3330,  -39,   20]])

Odpowiedź

       [[   2,    4, 1110,  -48,   10]
        [   6,    7, 1480,  -29,   10]
        [  22,   24, 4810,  -68,   30]
        [  16,   17, 3330,  -39,   20]]

# 4 Wprowadzenie do Pandas
## 4.1 Podstawy

**Pandas** - biblioteka do manipulacji ramkami danych, tak samo jak numpy jest szybka, lekka i wygodna. 

### Z poprzednich warsztatów
**Series**: Series jest to jednowymiarowa struktura danych („jednowymiarowa tablica”), która może przechowywać wartości i dla każdej wartości ma także unikalny indeks   
**DataFrame**: DataFrame to dwuwymiarowa (lub więcej) wymiarowa struktura danych - w zasadzie tabela z wierszami i kolumnami. Kolumny mają nazwy, a wiersze indeksy.

In [104]:
import pandas as pd 

In [105]:
# Import danych dzięki funkci pd.read_csv()
df = pd.read_csv("data/vgsales.csv", delimiter= ",").dropna().reset_index(drop=True) # na windows data/vgsales.csv

In [106]:
type(df)

pandas.core.frame.DataFrame

In [107]:
# spójrzmy na "głowę" naszej ramki danych
df.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [108]:
# pomijamy rank 
df = df.drop("Rank", axis = 1)

In [109]:
# gdy wywołamy zmienną ukaże się nam cała tabela, gdy będzie duża pokaże początek i koniec tabeli
df

Name Platform    Year  \
0                                            Wii Sports      Wii  2006.0   
1                                     Super Mario Bros.      NES  1985.0   
2                                        Mario Kart Wii      Wii  2008.0   
3                                     Wii Sports Resort      Wii  2009.0   
4                              Pokemon Red/Pokemon Blue       GB  1996.0   
...                                                 ...      ...     ...   
16286                Woody Woodpecker in Crazy Castle 5      GBA  2002.0   
16287                     Men in Black II: Alien Escape       GC  2003.0   
16288  SCORE International Baja 1000: The Official Game      PS2  2008.0   
16289                                        Know How 2       DS  2010.0   
16290                                  Spirits & Spells      GBA  2003.0   

              Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0            Sports    Nintendo     41.49     29.02      3.77         8.46   
1          Platform    Nintendo     29.08      3.58      6.81         0.77   
2            Racing    Nintendo     15.85     12.88      3.79         3.31   
3            Sports    Nintendo     15.75     11.01      3.28         2.96   
4      Role-Playing    Nintendo     11.27      8.89     10.22         1.00   
...             ...         ...       ...       ...       ...          ...   
16286      Platform       Kemco      0.01      0.00      0.00         0.00   
16287       Shooter  Infogrames      0.01      0.00      0.00         0.00   
16288        Racing  Activision      0.00      0.00      0.00         0.00   
16289        Puzzle    7G//AMES      0.00      0.01      0.00         0.00   
16290      Platform     Wanadoo      0.01      0.00      0.00         0.00   

       Global_Sales  
0             82.74  
1             40.24  
2             35.82  
3             33.00  
4             31.37  
...             ...  
16286          0.01  
16287          0.01  
16288          0.01  
16289          0.01  
16290          0.01  

[16291 rows x 10 columns]

### 4.1.1 Atrybuty pd.DataFrame

In [110]:
df.index

RangeIndex(start=0, stop=16291, step=1)

In [111]:
type(df.index)

pandas.core.indexes.range.RangeIndex

In [112]:
df.columns

Index(['Name', 'Platform', 'Year', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales'],
      dtype='object')

In [113]:
df.dtypes   # Sposób zapisu może się różnić od sposobu zapisu danych itp.

Name             object
Platform         object
Year            float64
Genre            object
Publisher        object
NA_Sales        float64
EU_Sales        float64
JP_Sales        float64
Other_Sales     float64
Global_Sales    float64
dtype: object

In [114]:
df.axes

[RangeIndex(start=0, stop=16291, step=1),
 Index(['Name', 'Platform', 'Year', 'Genre', 'Publisher', 'NA_Sales',
        'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales'],
       dtype='object')]

In [115]:
df.shape

(16291, 10)

### 4.1.2 Wybieranie kolumn, oraz czym jest kolumna

In [116]:
df["Platform"]  

0        Wii
1        NES
2        Wii
3        Wii
4         GB
        ... 
16286    GBA
16287     GC
16288    PS2
16289     DS
16290    GBA
Name: Platform, Length: 16291, dtype: object

In [117]:
# alternatywnie df.Platform - kiedy uzywac? 
df.JP_Sales

0         3.77
1         6.81
2         3.79
3         3.28
4        10.22
         ...  
16286     0.00
16287     0.00
16288     0.00
16289     0.00
16290     0.00
Name: JP_Sales, Length: 16291, dtype: float64

In [118]:
df[["Platform","Name"]]

Platform                                              Name
0          Wii                                        Wii Sports
1          NES                                 Super Mario Bros.
2          Wii                                    Mario Kart Wii
3          Wii                                 Wii Sports Resort
4           GB                          Pokemon Red/Pokemon Blue
...        ...                                               ...
16286      GBA                Woody Woodpecker in Crazy Castle 5
16287       GC                     Men in Black II: Alien Escape
16288      PS2  SCORE International Baja 1000: The Official Game
16289       DS                                        Know How 2
16290      GBA                                  Spirits & Spells

[16291 rows x 2 columns]

In [119]:
type(df.Platform)

pandas.core.series.Series



## 4.2 Indeksowanie

Omówimy 3 główne sposoby indeksowania w Pandas: 
1. []
2. .iloc[]
3. .loc[]

Dla lepszego wizualnego zrozumienia, początkowo wybierzemy kilka losowych wierszy

In [120]:
np.random.seed(123)
smalldf = df.iloc[np.random.randint(0,df.shape[0],10),:].reset_index(drop=True)

In [121]:
smalldf

Name Platform    Year     Genre  \
0             The Godfather (old US sales)       XB  2006.0    Action   
1  Nobunaga no Yabou Online: Houou no Shou      PS3  2012.0  Strategy   
2         Romance of the Three Kingdoms XI      PS2  2007.0  Strategy   
3                        Sonic Generations     X360  2011.0  Platform   
4                                     Vexx       GC  2003.0  Platform   
5                    Angry Birds Star Wars      Wii  2013.0  Strategy   
6       Saru! Get You! SaruSaru Daisakusen      PSP  2007.0  Platform   
7       Barnyard Blast: Swine of the Night       DS  2008.0  Platform   
8                   God Eater Resurrection      PSV  2015.0    Action   
9                      Star Trek: The Game       PC  2013.0    Action   

                     Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0              Electronic Arts      0.42      0.11      0.00         0.02   
1                  Ackkstudios      0.00      0.00      0.01         0.00   
2                   Tecmo Koei      0.00      0.00      0.07         0.00   
3                         Sega      0.71      0.57      0.00         0.14   
4        Acclaim Entertainment      0.05      0.01      0.00         0.00   
5                   Activision      0.21      0.10      0.00         0.03   
6  Sony Computer Entertainment      0.00      0.00      0.03         0.00   
7          Big Ben Interactive      0.02      0.00      0.00         0.00   
8           Namco Bandai Games      0.00      0.00      0.19         0.00   
9           Namco Bandai Games      0.00      0.02      0.00         0.00   

   Global_Sales  
0          0.55  
1          0.01  
2          0.07  
3          1.42  
4          0.06  
5          0.35  
6          0.03  
7          0.02  
8          0.19  
9          0.03

### Sposób 1) []

In [122]:
smalldf[1:3]

Name Platform    Year     Genre  \
1  Nobunaga no Yabou Online: Houou no Shou      PS3  2012.0  Strategy   
2         Romance of the Three Kingdoms XI      PS2  2007.0  Strategy   

     Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
1  Ackkstudios       0.0       0.0      0.01          0.0          0.01  
2   Tecmo Koei       0.0       0.0      0.07          0.0          0.07

In [123]:
# jak w numpy i w python
smalldf[0::2]  # - dla obserwacji

Name Platform    Year     Genre  \
0        The Godfather (old US sales)       XB  2006.0    Action   
2    Romance of the Three Kingdoms XI      PS2  2007.0  Strategy   
4                                Vexx       GC  2003.0  Platform   
6  Saru! Get You! SaruSaru Daisakusen      PSP  2007.0  Platform   
8              God Eater Resurrection      PSV  2015.0    Action   

                     Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0              Electronic Arts      0.42      0.11      0.00         0.02   
2                   Tecmo Koei      0.00      0.00      0.07         0.00   
4        Acclaim Entertainment      0.05      0.01      0.00         0.00   
6  Sony Computer Entertainment      0.00      0.00      0.03         0.00   
8           Namco Bandai Games      0.00      0.00      0.19         0.00   

   Global_Sales  
0          0.55  
2          0.07  
4          0.06  
6          0.03  
8          0.19

In [124]:
# nie można łączyć 
smalldf[0:4,1]

TypeError: '(slice(0, 4, None), 1)' is an invalid key

### Sposób 2) .loc[]

In [125]:
smalldf.loc[0:4,["Name"]]

Name
0             The Godfather (old US sales)
1  Nobunaga no Yabou Online: Houou no Shou
2         Romance of the Three Kingdoms XI
3                        Sonic Generations
4                                     Vexx

In [126]:
smalldf.columns[0:4]

Index(['Name', 'Platform', 'Year', 'Genre'], dtype='object')

In [127]:
# dzięki .loc[] możemy łączyć kolumny i indeksy, co jeszcze możemy robić? 
smalldf.loc[0:4, smalldf.columns[0:4]]

Name Platform    Year     Genre
0             The Godfather (old US sales)       XB  2006.0    Action
1  Nobunaga no Yabou Online: Houou no Shou      PS3  2012.0  Strategy
2         Romance of the Three Kingdoms XI      PS2  2007.0  Strategy
3                        Sonic Generations     X360  2011.0  Platform
4                                     Vexx       GC  2003.0  Platform

In [128]:
smalldf

Name Platform    Year     Genre  \
0             The Godfather (old US sales)       XB  2006.0    Action   
1  Nobunaga no Yabou Online: Houou no Shou      PS3  2012.0  Strategy   
2         Romance of the Three Kingdoms XI      PS2  2007.0  Strategy   
3                        Sonic Generations     X360  2011.0  Platform   
4                                     Vexx       GC  2003.0  Platform   
5                    Angry Birds Star Wars      Wii  2013.0  Strategy   
6       Saru! Get You! SaruSaru Daisakusen      PSP  2007.0  Platform   
7       Barnyard Blast: Swine of the Night       DS  2008.0  Platform   
8                   God Eater Resurrection      PSV  2015.0    Action   
9                      Star Trek: The Game       PC  2013.0    Action   

                     Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0              Electronic Arts      0.42      0.11      0.00         0.02   
1                  Ackkstudios      0.00      0.00      0.01         0.00   
2                   Tecmo Koei      0.00      0.00      0.07         0.00   
3                         Sega      0.71      0.57      0.00         0.14   
4        Acclaim Entertainment      0.05      0.01      0.00         0.00   
5                   Activision      0.21      0.10      0.00         0.03   
6  Sony Computer Entertainment      0.00      0.00      0.03         0.00   
7          Big Ben Interactive      0.02      0.00      0.00         0.00   
8           Namco Bandai Games      0.00      0.00      0.19         0.00   
9           Namco Bandai Games      0.00      0.02      0.00         0.00   

   Global_Sales  
0          0.55  
1          0.01  
2          0.07  
3          1.42  
4          0.06  
5          0.35  
6          0.03  
7          0.02  
8          0.19  
9          0.03

In [129]:
# możemy też wybierać "logicznie"
smalldf.loc[smalldf.Global_Sales > 1]

Name Platform    Year     Genre Publisher  NA_Sales  EU_Sales  \
3  Sonic Generations     X360  2011.0  Platform      Sega      0.71      0.57   

   JP_Sales  Other_Sales  Global_Sales  
3       0.0         0.14          1.42

In [130]:
smalldf.Global_Sales > 1

0    False
1    False
2    False
3     True
4    False
5    False
6    False
7    False
8    False
9    False
Name: Global_Sales, dtype: bool

In [131]:
smalldf.loc[(smalldf.Genre == "Platform") & (smalldf.Year < 2010)]

Name Platform    Year     Genre  \
4                                Vexx       GC  2003.0  Platform   
6  Saru! Get You! SaruSaru Daisakusen      PSP  2007.0  Platform   
7  Barnyard Blast: Swine of the Night       DS  2008.0  Platform   

                     Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
4        Acclaim Entertainment      0.05      0.01      0.00          0.0   
6  Sony Computer Entertainment      0.00      0.00      0.03          0.0   
7          Big Ben Interactive      0.02      0.00      0.00          0.0   

   Global_Sales  
4          0.06  
6          0.03  
7          0.02

### Sposób 3) .iloc[]

In [132]:
# dzieli .iloc[] możemy wybierać indeksy 
smalldf.iloc[0:3, 3:5]

Genre        Publisher
0    Action  Electronic Arts
1  Strategy      Ackkstudios
2  Strategy       Tecmo Koei

In [133]:
smalldf.iloc[[0,1,2],[3,4]]

Genre        Publisher
0    Action  Electronic Arts
1  Strategy      Ackkstudios
2  Strategy       Tecmo Koei

In [134]:
smalldf.iloc[2:10:3,2:10:2]

Year           Publisher  EU_Sales  Other_Sales
2  2007.0          Tecmo Koei       0.0         0.00
5  2013.0          Activision       0.1         0.03
8  2015.0  Namco Bandai Games       0.0         0.00

### Podsumowując 
* Gdy chcemy kolumnę po nazwie -> []
* Gdy chcemy obserwacje, kolumny po wartościach logicznych lub ich nazwach -> .loc[]
* Gdy chcemy obserwacje, kolumny po ich indeksach -> .iloc[]

## 4.3 Metody ramek danych

Ramki danych jako obiekty mają wiele przydatnych metod. Poznamy niektóre z nich.

In [135]:
df.Name

0                                              Wii Sports
1                                       Super Mario Bros.
2                                          Mario Kart Wii
3                                       Wii Sports Resort
4                                Pokemon Red/Pokemon Blue
                               ...                       
16286                  Woody Woodpecker in Crazy Castle 5
16287                       Men in Black II: Alien Escape
16288    SCORE International Baja 1000: The Official Game
16289                                          Know How 2
16290                                    Spirits & Spells
Name: Name, Length: 16291, dtype: object

In [136]:
df.Name.count()

16291

In [137]:
df.Name.unique()

array(['Wii Sports', 'Super Mario Bros.', 'Mario Kart Wii', ...,
       'Plushees', 'Woody Woodpecker in Crazy Castle 5', 'Know How 2'],
      dtype=object)

In [138]:
df.Name.unique().shape

(11325,)

In [139]:
df.Year.mode()

0    2009.0
dtype: float64

In [140]:
df.Global_Sales.median()

0.17

In [141]:
df.Global_Sales.max()

82.74

In [142]:
df.describe()

Year      NA_Sales      EU_Sales      JP_Sales   Other_Sales  \
count  16291.000000  16291.000000  16291.000000  16291.000000  16291.000000   
mean    2006.405561      0.265647      0.147731      0.078833      0.048426   
std        5.832412      0.822432      0.509303      0.311879      0.190083   
min     1980.000000      0.000000      0.000000      0.000000      0.000000   
25%     2003.000000      0.000000      0.000000      0.000000      0.000000   
50%     2007.000000      0.080000      0.020000      0.000000      0.010000   
75%     2010.000000      0.240000      0.110000      0.040000      0.040000   
max     2020.000000     41.490000     29.020000     10.220000     10.570000   

       Global_Sales  
count  16291.000000  
mean       0.540910  
std        1.567345  
min        0.010000  
25%        0.060000  
50%        0.170000  
75%        0.480000  
max       82.740000

In [143]:
# Ile gier sprzedalo sie w nakładzie wiekszym niż 1 mln
df.loc[df.Global_Sales > 1].shape

(2031, 10)

In [144]:
# Jaki to był procent wszystkich gier
df.Global_Sales[df.Global_Sales > 1].count()/df.Global_Sales.count() * 100

12.467006322509361

In [145]:
df.Year.min()

1980.0

In [146]:
year_min = df.Year.argmin()
year_min

257

In [147]:
df.Year[year_min]

1980.0

In [148]:
df.NA_Sales + df.EU_Sales

0        70.51
1        32.66
2        28.73
3        26.76
4        20.16
         ...  
16286     0.01
16287     0.01
16288     0.00
16289     0.01
16290     0.01
Length: 16291, dtype: float64

In [149]:
df.EU_Sales.sort_values(ascending = False)

0        29.02
2        12.88
3        11.01
10       11.00
16        9.27
         ...  
10308     0.00
10306     0.00
10303     0.00
10297     0.00
8145      0.00
Name: EU_Sales, Length: 16291, dtype: float64

In [150]:
df.sort_values("EU_Sales", ascending = False).head()

Name Platform    Year       Genre             Publisher  \
0           Wii Sports      Wii  2006.0      Sports              Nintendo   
2       Mario Kart Wii      Wii  2008.0      Racing              Nintendo   
3    Wii Sports Resort      Wii  2009.0      Sports              Nintendo   
10          Nintendogs       DS  2005.0  Simulation              Nintendo   
16  Grand Theft Auto V      PS3  2013.0      Action  Take-Two Interactive   

    NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0      41.49     29.02      3.77         8.46         82.74  
2      15.85     12.88      3.79         3.31         35.82  
3      15.75     11.01      3.28         2.96         33.00  
10      9.07     11.00      1.93         2.75         24.76  
16      7.01      9.27      0.97         4.14         21.40

In [151]:
df.sort_values("EU_Sales", ascending = False).reset_index().head()

index                Name Platform    Year       Genre  \
0      0          Wii Sports      Wii  2006.0      Sports   
1      2      Mario Kart Wii      Wii  2008.0      Racing   
2      3   Wii Sports Resort      Wii  2009.0      Sports   
3     10          Nintendogs       DS  2005.0  Simulation   
4     16  Grand Theft Auto V      PS3  2013.0      Action   

              Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0              Nintendo     41.49     29.02      3.77         8.46   
1              Nintendo     15.85     12.88      3.79         3.31   
2              Nintendo     15.75     11.01      3.28         2.96   
3              Nintendo      9.07     11.00      1.93         2.75   
4  Take-Two Interactive      7.01      9.27      0.97         4.14   

   Global_Sales  
0         82.74  
1         35.82  
2         33.00  
3         24.76  
4         21.40

In [152]:
df.sort_values("EU_Sales", ascending = False).reset_index(drop=True).head()

Name Platform    Year       Genre             Publisher  \
0          Wii Sports      Wii  2006.0      Sports              Nintendo   
1      Mario Kart Wii      Wii  2008.0      Racing              Nintendo   
2   Wii Sports Resort      Wii  2009.0      Sports              Nintendo   
3          Nintendogs       DS  2005.0  Simulation              Nintendo   
4  Grand Theft Auto V      PS3  2013.0      Action  Take-Two Interactive   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     15.85     12.88      3.79         3.31         35.82  
2     15.75     11.01      3.28         2.96         33.00  
3      9.07     11.00      1.93         2.75         24.76  
4      7.01      9.27      0.97         4.14         21.40

In [153]:
df.sort_values(["Genre", "Platform"]).reset_index(drop=True).head()

Name Platform    Year   Genre  \
0                      Frogger     2600  1981.0  Action   
1  E.T.: The Extra Terrestrial     2600  1981.0  Action   
2                   Spider-Man     2600  1981.0  Action   
3             Custer's Revenge     2600  1981.0  Action   
4                        Alien     2600  1981.0  Action   

                      Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0                  Parker Bros.      2.06      0.12       0.0         0.02   
1                         Atari      1.84      0.11       0.0         0.02   
2                  Parker Bros.      0.87      0.05       0.0         0.01   
3                      Mystique      0.76      0.05       0.0         0.01   
4  20th Century Fox Video Games      0.74      0.04       0.0         0.01   

   Global_Sales  
0          2.20  
1          1.97  
2          0.93  
3          0.82  
4          0.79

## 4.4 Zadanie  (5 min)

### Zadanie 1
Wybierz te gry, które sprzedały się lepiej W Europie niż w Ameryce Północnej. Ile ich było? Jaki to procent wszystkich gier? 

## 4.5 Groupby & Merge

![groupby](img/groupby.png)

Illustration from book : Python for Data Analysis. Data Wrangling with Pandas, Numpy and Ipython (2017, O'Reilly)

In [190]:
df.groupby("Genre")

In [196]:
df.groupby("Genre").Name

In [194]:
df.groupby("Genre")[["Name", "Global_Sales"]]

In [197]:
df.groupby("Genre").Name.agg("count")

Genre
Action          3251
Adventure       1274
Fighting         836
Misc            1686
Platform         875
Puzzle           570
Racing          1225
Role-Playing    1470
Shooter         1282
Simulation       848
Sports          2304
Strategy         670
Name: Name, dtype: int64

In [198]:
df.groupby("Genre").Global_Sales

In [200]:
df.groupby("Genre").Global_Sales.agg("sum")

Genre
Action          1722.84
Adventure        234.59
Fighting         444.05
Misc             789.87
Platform         829.13
Puzzle           242.21
Racing           726.76
Role-Playing     923.83
Shooter         1026.20
Simulation       389.98
Sports          1309.24
Strategy         173.27
Name: Global_Sales, dtype: float64

In [204]:
df.groupby("Genre").Global_Sales.agg("max")

Genre
Action          21.40
Adventure       11.18
Fighting        13.04
Misc            29.02
Platform        40.24
Puzzle          30.26
Racing          35.82
Role-Playing    31.37
Shooter         28.31
Simulation      24.76
Sports          82.74
Strategy         5.45
Name: Global_Sales, dtype: float64

**grupować można nie tylko po jednej kolumnie** 



In [209]:
df2 = df.groupby(["Genre","Platform"])[["Global_Sales", "Name"]].agg("max")

In [210]:
df2

Global_Sales                                           Name
Genre    Platform                                                             
Action   2600              2.20                                          X-Man
         3DS               4.58  Zyuden Sentai Kyoryuger: Game de Gaburincho!!
         DC                1.14                 Resident Evil - Code: Veronica
         DS               11.90                                  Zombi Daisuki
         GB                3.83            The Legend of Zelda: Oracle of Ages
...                         ...                                            ...
Strategy Wii               0.66                   Yu-Gi-Oh! 5D's: Duel Transer
         WiiU              1.11              Romance of the Three Kingdoms XII
         X360              2.63                                  Zoids Assault
         XB                0.73                  Yu-Gi-Oh! The Dawn of Destiny
         XOne              0.20                                         XCOM 2

[293 rows x 2 columns]

In [165]:
# a co jeśli chcemy wnioski z naszej wynikowej tabeli zawrzeć do tej pierwotnej? 

In [212]:
df3 = df.groupby("Platform").Global_Sales.agg("sum")
df3 = df3.to_frame()
df3 = df3.reset_index()

In [213]:
df3

Platform  Global_Sales
0      2600         86.57
1       3DO          0.10
2       3DS        246.27
3        DC         15.97
4        DS        818.91
5        GB        254.42
6       GBA        305.62
7        GC        197.14
8       GEN         28.36
9        GG          0.04
10      N64        218.21
11      NES        251.07
12       NG          1.44
13       PC        254.70
14     PCFX          0.03
15       PS        727.39
16      PS2       1233.46
17      PS3        949.35
18      PS4        278.10
19      PSP        291.71
20      PSV         61.60
21      SAT         33.59
22      SCD          1.87
23     SNES        200.05
24     TG16          0.16
25       WS          1.42
26      Wii        909.81
27     WiiU         81.86
28     X360        969.60
29       XB        252.09
30     XOne        141.06

In [214]:
df3.shape

(31, 2)

In [215]:
df3.head()


Platform  Global_Sales
0     2600         86.57
1      3DO          0.10
2      3DS        246.27
3       DC         15.97
4       DS        818.91

In [169]:
# Teraz wyniki naszej analizy połączymy z naszą tabelką df2

In [220]:
df4 = df.merge(df3, on = "Platform")
df4

Name Platform    Year         Genre  \
0                                Wii Sports      Wii  2006.0        Sports   
1                            Mario Kart Wii      Wii  2008.0        Racing   
2                         Wii Sports Resort      Wii  2009.0        Sports   
3                                  Wii Play      Wii  2006.0          Misc   
4                 New Super Mario Bros. Wii      Wii  2009.0      Platform   
...                                     ...      ...     ...           ...   
16286                           Policenauts      3DO  1995.0     Adventure   
16287                           Bust-A-Move      3DO  1994.0        Puzzle   
16288  Sotsugyou II: Neo Generation Special      3DO  1995.0    Simulation   
16289          Sonic the Hedgehog 2 (8-bit)       GG  1992.0      Platform   
16290    Blue Breaker: Ken Yorimo Hohoemi o     PCFX  1996.0  Role-Playing   

                          Publisher  NA_Sales  EU_Sales  JP_Sales  \
0                          Nintendo     41.49     29.02      3.77   
1                          Nintendo     15.85     12.88      3.79   
2                          Nintendo     15.75     11.01      3.28   
3                          Nintendo     14.03      9.20      2.93   
4                          Nintendo     14.59      7.06      4.70   
...                             ...       ...       ...       ...   
16286  Konami Digital Entertainment      0.00      0.00      0.06   
16287                   Micro Cabin      0.00      0.00      0.02   
16288                    Imageworks      0.00      0.00      0.02   
16289                          Sega      0.00      0.00      0.04   
16290                           NEC      0.00      0.00      0.03   

       Other_Sales  Global_Sales_x  Global_Sales_y  
0             8.46           82.74          909.81  
1             3.31           35.82          909.81  
2             2.96           33.00          909.81  
3             2.85           29.02          909.81  
4             2.26           28.62          909.81  
...            ...             ...             ...  
16286         0.00            0.06            0.10  
16287         0.00            0.02            0.10  
16288         0.00            0.02            0.10  
16289         0.00            0.04            0.04  
16290         0.00            0.03            0.03  

[16291 rows x 11 columns]

In [222]:
df4.rename(columns = {'Global_Sales_y':"Platform Global Sales", 
                      'Global_Sales_x': "Global_Sales"})

Name Platform    Year         Genre  \
0                                Wii Sports      Wii  2006.0        Sports   
1                            Mario Kart Wii      Wii  2008.0        Racing   
2                         Wii Sports Resort      Wii  2009.0        Sports   
3                                  Wii Play      Wii  2006.0          Misc   
4                 New Super Mario Bros. Wii      Wii  2009.0      Platform   
...                                     ...      ...     ...           ...   
16286                           Policenauts      3DO  1995.0     Adventure   
16287                           Bust-A-Move      3DO  1994.0        Puzzle   
16288  Sotsugyou II: Neo Generation Special      3DO  1995.0    Simulation   
16289          Sonic the Hedgehog 2 (8-bit)       GG  1992.0      Platform   
16290    Blue Breaker: Ken Yorimo Hohoemi o     PCFX  1996.0  Role-Playing   

                          Publisher  NA_Sales  EU_Sales  JP_Sales  \
0                          Nintendo     41.49     29.02      3.77   
1                          Nintendo     15.85     12.88      3.79   
2                          Nintendo     15.75     11.01      3.28   
3                          Nintendo     14.03      9.20      2.93   
4                          Nintendo     14.59      7.06      4.70   
...                             ...       ...       ...       ...   
16286  Konami Digital Entertainment      0.00      0.00      0.06   
16287                   Micro Cabin      0.00      0.00      0.02   
16288                    Imageworks      0.00      0.00      0.02   
16289                          Sega      0.00      0.00      0.04   
16290                           NEC      0.00      0.00      0.03   

       Other_Sales  Global_Sales  Platform Global Sales  
0             8.46         82.74                 909.81  
1             3.31         35.82                 909.81  
2             2.96         33.00                 909.81  
3             2.85         29.02                 909.81  
4             2.26         28.62                 909.81  
...            ...           ...                    ...  
16286         0.00          0.06                   0.10  
16287         0.00          0.02                   0.10  
16288         0.00          0.02                   0.10  
16289         0.00          0.04                   0.04  
16290         0.00          0.03                   0.03  

[16291 rows x 11 columns]

In [223]:
df4.head() # ważne żeby przypisać

Name Platform    Year     Genre Publisher  NA_Sales  \
0                 Wii Sports      Wii  2006.0    Sports  Nintendo     41.49   
1             Mario Kart Wii      Wii  2008.0    Racing  Nintendo     15.85   
2          Wii Sports Resort      Wii  2009.0    Sports  Nintendo     15.75   
3                   Wii Play      Wii  2006.0      Misc  Nintendo     14.03   
4  New Super Mario Bros. Wii      Wii  2009.0  Platform  Nintendo     14.59   

   EU_Sales  JP_Sales  Other_Sales  Global_Sales_x  Global_Sales_y  
0     29.02      3.77         8.46           82.74          909.81  
1     12.88      3.79         3.31           35.82          909.81  
2     11.01      3.28         2.96           33.00          909.81  
3      9.20      2.93         2.85           29.02          909.81  
4      7.06      4.70         2.26           28.62          909.81

In [224]:
df4 = df4.rename(columns = {'Global_Sales_y':"Aggregated Global Sales", 
                            'Global_Sales_x': "Global_Sales"})

In [226]:
df4.head()

Name Platform    Year     Genre Publisher  NA_Sales  \
0                 Wii Sports      Wii  2006.0    Sports  Nintendo     41.49   
1             Mario Kart Wii      Wii  2008.0    Racing  Nintendo     15.85   
2          Wii Sports Resort      Wii  2009.0    Sports  Nintendo     15.75   
3                   Wii Play      Wii  2006.0      Misc  Nintendo     14.03   
4  New Super Mario Bros. Wii      Wii  2009.0  Platform  Nintendo     14.59   

   EU_Sales  JP_Sales  Other_Sales  Global_Sales  Aggregated Global Sales  
0     29.02      3.77         8.46         82.74                   909.81  
1     12.88      3.79         3.31         35.82                   909.81  
2     11.01      3.28         2.96         33.00                   909.81  
3      9.20      2.93         2.85         29.02                   909.81  
4      7.06      4.70         2.26         28.62                   909.81

## 4.6 Zastosowania

### Sprawdźmy gry które sprzedały się najlepiej w swojej kategorii

In [175]:
# najpierw przekształcimy delikatnie nasz data frame

In [227]:
df.head()

Name Platform    Year         Genre Publisher  \
0                Wii Sports      Wii  2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [228]:
best_games = df.groupby("Genre")["Global_Sales"].max()
best_games

Genre
Action          21.40
Adventure       11.18
Fighting        13.04
Misc            29.02
Platform        40.24
Puzzle          30.26
Racing          35.82
Role-Playing    31.37
Shooter         28.31
Simulation      24.76
Sports          82.74
Strategy         5.45
Name: Global_Sales, dtype: float64

In [229]:
df.loc[df.Global_Sales.isin(best_games), :]

Name Platform    Year         Genre  \
0                            Wii Sports      Wii  2006.0        Sports   
1                     Super Mario Bros.      NES  1985.0      Platform   
2                        Mario Kart Wii      Wii  2008.0        Racing   
4              Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing   
5                                Tetris       GB  1989.0        Puzzle   
7                              Wii Play      Wii  2006.0          Misc   
9                             Duck Hunt      NES  1984.0       Shooter   
10                           Nintendogs       DS  2005.0    Simulation   
16                   Grand Theft Auto V      PS3  2013.0        Action   
39              Super Smash Bros. Brawl      Wii  2008.0      Fighting   
50   Super Mario Land 2: 6 Golden Coins       GB  1992.0     Adventure   
165                     Pokemon Stadium      N64  1999.0      Strategy   

                Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0                Nintendo     41.49     29.02      3.77         8.46   
1                Nintendo     29.08      3.58      6.81         0.77   
2                Nintendo     15.85     12.88      3.79         3.31   
4                Nintendo     11.27      8.89     10.22         1.00   
5                Nintendo     23.20      2.26      4.22         0.58   
7                Nintendo     14.03      9.20      2.93         2.85   
9                Nintendo     26.93      0.63      0.28         0.47   
10               Nintendo      9.07     11.00      1.93         2.75   
16   Take-Two Interactive      7.01      9.27      0.97         4.14   
39               Nintendo      6.75      2.61      2.66         1.02   
50               Nintendo      6.16      2.04      2.69         0.29   
165              Nintendo      3.18      1.24      0.94         0.09   

     Global_Sales  
0           82.74  
1           40.24  
2           35.82  
4           31.37  
5           30.26  
7           29.02  
9           28.31  
10          24.76  
16          21.40  
39          13.04  
50          11.18  
165          5.45

Bardziej skomplikowane zadanie 
### Sprawdzimy jaka była sprzedaż gier na konsole w 2 roku ich sprzedaży. Pokaż tylko te konsole w których sprzedaż w 2 roku była mniejsza niż w 1. 

In [230]:
start_year = df.groupby("Platform").Year.min()
start_year[:5]

Platform
2600    1980.0
3DO     1994.0
3DS     2011.0
DC      1998.0
DS      1985.0
Name: Year, dtype: float64

In [231]:
df5 = df.groupby(["Platform","Year"]).Global_Sales.sum().to_frame()
df5.head()

Global_Sales
Platform Year                
2600     1980.0         11.38
         1981.0         35.77
         1982.0         28.86
         1983.0          5.83
         1984.0          0.27

In [232]:
df5 = df5.reset_index()
df5

Platform    Year  Global_Sales
0       2600  1980.0         11.38
1       2600  1981.0         35.77
2       2600  1982.0         28.86
3       2600  1983.0          5.83
4       2600  1984.0          0.27
..       ...     ...           ...
236       XB  2008.0          0.18
237     XOne  2013.0         18.60
238     XOne  2014.0         52.43
239     XOne  2015.0         57.66
240     XOne  2016.0         12.37

[241 rows x 3 columns]

In [182]:
# teraz chcemy jakoś zamarkować rok konsoli na rynku jako pierwszy lub drugi

In [234]:
start_year = start_year.to_frame().reset_index()
start_year[:5]

Platform    Year
0     2600  1980.0
1      3DO  1994.0
2      3DS  2011.0
3       DC  1998.0
4       DS  1985.0

In [236]:
start_year = start_year.rename(columns = {"Year":"First Year"})
start_year.head()

Platform  First Year
0     2600      1980.0
1      3DO      1994.0
2      3DS      2011.0
3       DC      1998.0
4       DS      1985.0

In [237]:
df5 = df5.merge(start_year, on="Platform")
df5

Platform    Year  Global_Sales  First Year
0       2600  1980.0         11.38      1980.0
1       2600  1981.0         35.77      1980.0
2       2600  1982.0         28.86      1980.0
3       2600  1983.0          5.83      1980.0
4       2600  1984.0          0.27      1980.0
..       ...     ...           ...         ...
236       XB  2008.0          0.18      2000.0
237     XOne  2013.0         18.60      2013.0
238     XOne  2014.0         52.43      2013.0
239     XOne  2015.0         57.66      2013.0
240     XOne  2016.0         12.37      2013.0

[241 rows x 4 columns]

In [238]:
first_year_sales = df5.loc[(df5.Year == df5["First Year"])]

In [239]:
second_year_sales = df5.loc[df5.Year == df5["First Year"] + 1]

In [240]:
end_df = first_year_sales.merge(second_year_sales, on = ["Platform"])
end_df.head()

Platform  Year_x  Global_Sales_x  First Year_x  Year_y  Global_Sales_y  \
0     2600  1980.0           11.38        1980.0  1981.0           35.77   
1      3DO  1994.0            0.02        1994.0  1995.0            0.08   
2      3DS  2011.0           62.53        2011.0  2012.0           51.14   
3       DC  1998.0            3.39        1998.0  1999.0            5.17   
4       GB  1988.0            1.43        1988.0  1989.0           64.98   

   First Year_y  
0        1980.0  
1        1994.0  
2        2011.0  
3        1998.0  
4        1988.0

In [189]:
end_df.loc[end_df.Global_Sales_x > end_df.Global_Sales_y]

Platform  Year_x  Global_Sales_x  First Year_x  Year_y  Global_Sales_y  \
2       3DS  2011.0           62.53        2011.0  2012.0           51.14   
18      SCD  1993.0            1.50        1993.0  1994.0            0.37   
19     SNES  1990.0           26.16        1990.0  1991.0           16.21   

    First Year_y  
2         2011.0  
18        1993.0  
19        1990.0

## 4.7 Zadania (Do domu ) (15 min)

### Zadanie 1

Stwórz ramkę danych w której uwzględnisz gry na PS3 (`platform`), które sprzedały się lepiej w Europie niż w Ameryce

### Zadanie 2

Odpowiedz na pytanie: 
Jakie typy gier sprzedają się sumarycznie lepiej w Japonii niż w pozostałych obszarach świata? 